ethcore/sync/src/chain/fork_filter.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+// Re-export eip-2124 crate contents here.
+pub use eip_2124::{BlockNumber, ForkId, RejectReason};
+
+use client_traits::BlockChainClient;
+use eip_2124::ForkFilter;
+use engine::Engine;
+use maplit::btreeset;
+use parity_util_mem::MallocSizeOf;
+
+/// Wrapper around fork filter that provides integration with `ForkFilter`.
+#[derive(MallocSizeOf)]
+pub struct ForkFilterApi {
+	inner: ForkFilter,
+}
+
+impl ForkFilterApi {
+	/// Create `ForkFilterApi` from `Engine` and a closure that returns `BlockChainInfo`
+	pub fn new(client: &dyn BlockChainClient, engine: &dyn Engine) -> Self {
+		let chain_info = client.chain_info();
+		let params = engine.params();
+		let mut forks = btreeset![
+			params.eip150_transition,
+			params.eip160_transition,
+			params.eip161abc_transition,
+			params.eip161d_transition,
+			params.eip98_transition,
+			params.eip658_transition,
+			params.eip155_transition,
+			params.validate_receipts_transition,
+			params.validate_chain_id_transition,
+			params.eip140_transition,
+			params.eip210_transition,
+			params.eip211_transition,
+			params.eip214_transition,
+			params.eip145_transition,
+			params.eip1052_transition,
+			params.eip1283_transition,
+			params.eip1283_disable_transition,
+			params.eip1283_reenable_transition,
+			params.eip1014_transition,
+			params.eip1706_transition,
+			params.eip1344_transition,
+			params.eip1884_transition,
+			params.eip2028_transition,
+			params.eip2200_advance_transition,
+			params.dust_protection_transition,
+			params.wasm_activation_transition,
+			params.kip4_transition,
+			params.kip6_transition,
+			params.max_code_size_transition,
+			params.transaction_permission_contract_transition,
+		];
+		// BUG: Rinkeby has homestead transition at block 1 but we can't reflect that in specs for non-Ethash networks
+		if params.network_id == 0x4 {
+			forks.insert(1);
+		}
+		// Specific transitions for Ethash-based networks
+		if let Some(ext) = engine.machine().ethash_extensions() {
+			forks.insert(ext.homestead_transition);
+			forks.insert(ext.dao_hardfork_transition);
+		}
+		// Ethereum's difficulty bomb delay is a fork too
+		if let Some(ethash_engine) = engine.downcast_ref::<ethash_engine::Ethash>() {
```

Comment:
So I guess this is the change we're most worried about. What options do we have here? Can we think of a hack that lets us get around depending on the engine here?